### ISSUE TITLE
unused_closure_parameter false positive on lazy var

### ISSUE THREAD

**Issue by ignazioc** [realm/SwiftLint#1161]
When a closure is used in this way, `swiftlint` raise a warning about the variable `lbl` not being used.

```swift
    lazy var label: UILabel = { (lbl: UILabel) -> UILabel in
        lbl.backgroundColor = UIColor.red
        return lbl
    }(UILabel())
```

**Comment by jpsim** [comment]
Thanks for reporting this.

**Comment by ignazioc** [comment]
It would be nice to have kind of a tutor to fix on my own :)

-- 
Ignazio Calo'

**Comment by jpsim** [comment]
Some tips to debug this in SwiftLint yourself:

1. Clone and update submodules.
1. Build and run the `swiftlint` scheme from `SwiftLint.xcworkspace`, setting the scheme's "Working Directory" and arguments.
1. Set a breakpoint in the rule you're trying to debug so you can step through an inspect the state as you step through to identify why the code you shared is triggering a violation.

![image](https://cloud.githubusercontent.com/assets/474794/21860039/c1c058ea-d7e3-11e6-9c11-86bd779fa157.png)

![image](https://cloud.githubusercontent.com/assets/474794/21860059/d513a41a-d7e3-11e6-86bf-85d538967fc4.png)

![image](https://cloud.githubusercontent.com/assets/474794/21860142/2f14f69e-d7e4-11e6-955a-98e30c66df3e.png)

**Comment by marcelofabri** [comment]
You can also add your example to the rule examples. Another useful thing might be updating the rule to conform to `OptInRule` so your breakpoints only will be triggered with the examples (and not all files from the integration tests)

**Comment by marcelofabri** [comment]
Another tip is using [SourceKitten](https://github.com/jpsim/SourceKitten) to get the information about your example so you can understand the problem better.

**Comment by ignazioc** [comment]
Thanks for the suggestions. I was able to debug the issue and investigate a little bit using `SourceKitten`. Here what I've found. 

This is the test input:
```swift

3.times { number in
	 number
}


{ ( lbl: UILabel) -> UILabel in
	lbl
}(obj)
```
`SwiftLint`  raises a warning on the second closure
> example.swift:7:5: warning: Unused Closure Parameter Violation: Unused parameter "lbl" in a closure should be replaced with _. (unused_closure_parameter)

Here the relevant output from SourceKitten:

```
"key.length" : 30,
"key.name" : "3.times",
"key.bodyoffset" : 10,
"key.kind" : "source.lang.swift.expr.call",
"key.offset" : 1,
"key.namelength" : 7

"key.length" : 43,
"key.name" : "{ ( lbl: UILabel) -> UILabel in\n\tlbl\n}",
"key.bodyoffset" : 73,
"key.kind" : "source.lang.swift.expr.call",
"key.offset" : 34,
"key.namelength" : 38
``` 

On the first example the closure's body can be found between `key.bodyoffset` and `key.lenghth` and `SwiftLint` code scans that part of the file to find any usage of the parameters.

Unfortunately on the second example the same indexes point outside of the definition of the closure, to the call (`(obj)`) part. Therefor `SwifLint` doesn't find any usage of the parameter and raise the (wrong) warning.

The first solution coming to my mind is to check if the `key.name` is actually a closure and in that case search there any usage of the parameter.
It doesn't look really clean so maybe some of you expert can give me an hint?

Thanks!

**Comment by marcelofabri** [comment]
I think it's hacky to try to use `key.name` because we don't have any AST information about its contents. What if the parameter is used in a string or comment? 

IMHO, we should just detect this case (i.e. detect that `key.name` is a closure) and skip any validations for now.

**Comment by marcelofabri** [comment]
After all, this rule doesn't work with closure declarations as well (see #1082).

**Comment by aamctustwo** [comment]
Not to hijack this thread, but a number of the suggestions given should probably be added to the `CONTRIBUTING.md` file to help future contributors.

**Comment by ignazioc** [comment]
agree.

**Comment by marcelofabri** [comment]
@ignazioc I've opened #1252 to ignore this case for now and we can deal with it later in #1082.
